File: test/progress.test.ts
import { describe, expect, it } from 'vitest';
import { isNdjsonProgressLine, isPureThinkingLine } from '../src/workers/progress.js';

describe('progress classifier (PRD Q33)', () => {
  describe('isPureThinkingLine', () => {
    it('returns true for thinking_delta lines', () => {
      expect(isPureThinkingLine('{"type":"message_update","assistantMessageEvent":{"type":"thinking_delta","delta":" = \""}}')).toBe(true);
    });
    it('returns true for thinking_start lines', () => {
      expect(isPureThinkingLine('{"type":"message_update","assistantMessageEvent":{"type":"thinking_start"}}')).toBe(true);
    });
    it('does not flag tool or text lines', () => {
      expect(isPureThinkingLine('{"type":"tool_execution_start","toolName":"bash"}')).toBe(false);
      expect(isPureThinkingLine('{"type":"message_update","assistantMessageEvent":{"type":"text_end","content":"hi"}}')).toBe(false);
    });
  });

  describe('isNdjsonProgressLine', () => {
    it('progress: tool_execution_start', () => {
      expect(isNdjsonProgressLine('{"type":"tool_execution_start","toolName":"read","args":{}}')).toBe(true);
    });
    it('progress: tool_execution_end', () => {
      expect(isNdjsonProgressLine('{"type":"tool_execution_end","toolName":"bash","result":{}}')).toBe(true);
    });
    it('progress: text_end (answer turn completed)', () => {
      expect(isNdjsonProgressLine('{"type":"message_update","assistantMessageEvent":{"type":"text_end","content":"hi"}}')).toBe(true);
    });
    it('progress: toolcall_start (tool arg assembled)', () => {
      expect(isNdjsonProgressLine('{"type":"message_update","assistantMessageEvent":{"type":"toolcall_start","id":"c1","toolName":"bash"}}')).toBe(true);
    });
    it('not progress: thinking_delta', () => {
      expect(isNdjsonProgressLine('{"type":"message_update","assistantMessageEvent":{"type":"thinking_delta","delta":" x"}}')).toBe(false);
    });
    it('not progress: empty line', () => {
      expect(isNdjsonProgressLine('')).toBe(false);
      expect(isNdjsonProgressLine('   ')).toBe(false);
    });
    it('not progress: arbitrary non-progress JSON', () => {
      expect(isNdjsonProgressLine('{"type":"message_update","assistantMessageEvent":{"type":"thinking_delta","delta":" hi"}}')).toBe(false);
      expect(isNdjsonProgressLine('{"type":"session","id":"abc"}')).toBe(false);
    });
  });
});

File: internal/workers/progress_test.go
// Go port of test/progress.test.ts (PRD Q33): the no-progress watchdog
// must not treat model deliberation as progress.
package workers

import "testing"

func TestIsPureThinkingLine(t *testing.T) {
	cases := []struct {
		name string
		line string
		want bool
	}{
		{"thinking_delta", `{"type":"message_update","assistantMessageEvent":{"type":"thinking_delta","delta":" = \""}}"`, true},
		{"thinking_start", `{"type":"message_update","assistantMessageEvent":{"type":"thinking_start"}}`, true},
		{"tool line not flagged", `{"type":"tool_execution_start","toolName":"bash"}`, false},
		{"text line not flagged", `{"type":"message_update","assistantMessageEvent":{"type":"text_end","content":"hi"}}`, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := IsPureThinkingLine(tc.line); got != tc.want {
				t.Fatalf("IsPureThinkingLine(%q) = %v, want %v", tc.line, got, tc.want)
			}
		})
	}
}

func TestIsNdjsonProgressLine(t *testing.T) {
	cases := []struct {
		name string
		line string
		want bool
	}{
		{"tool_execution_start", `{"type":"tool_execution_start","toolName":"read","args":{}}`, true},
		{"tool_execution_end", `{"type":"tool_execution_end","toolName":"bash","result":{}}`, true},
		{"text_end (answer turn completed)", `{"type":"message_update","assistantMessageEvent":{"type":"text_end","content":"hi"}}`, true},
		{"toolcall_start (tool arg assembled)", `{"type":"message_update","assistantMessageEvent":{"type":"toolcall_start","id":"c1","toolName":"bash"}}`, true},
		{"thinking_delta not progress", `{"type":"message_update","assistantMessageEvent":{"type":"thinking_delta","delta":" x"}}`, false},
		{"empty line", "", false},
		{"blank line", "   ", false},
		{"session header", `{"type":"session","id":"abc"}`, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := IsNdjsonProgressLine(tc.line); got != tc.want {
				t.Fatalf("IsNdjsonProgressLine(%q) = %v, want %v", tc.line, got, tc.want)
			}
		})
	}
}
